src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -31,49 +32,58 @@ public static class ActivatorUtilities
+            ConstructorInfo[] constructors = instanceType.GetConstructors();
+            if (constructors.Length == 0)
+            {
+                throw new InvalidOperationException(message);
+            }
 
-            if (!instanceType.IsAbstract)
+            ConstructorInfo? pereferredConstructor = null;
+            foreach (ConstructorInfo constructor in constructors)
             {
-                foreach (ConstructorInfo? constructor in instanceType.GetConstructors())
+                if (constructor.IsDefined(typeof(ActivatorUtilitiesConstructorAttribute)))
                 {
-                    var matcher = new ConstructorMatcher(constructor);
-                    bool isPreferred = constructor.IsDefined(typeof(ActivatorUtilitiesConstructorAttribute), false);
-                    int length = matcher.Match(parameters);
-
-                    if (isPreferred)
+                    if (pereferredConstructor is not null)
                     {
-                        if (seenPreferred)
-                        {
-                            ThrowMultipleCtorsMarkedWithAttributeException();
-                        }
-
-                        if (length == -1)
-                        {
-                            ThrowMarkedCtorDoesNotTakeAllProvidedArguments();
-                        }
-                    }
-
-                    if (isPreferred || bestLength < length)
-                    {
-                        bestLength = length;
-                        bestMatcher = matcher;
+                        ThrowMultipleCtorsMarkedWithAttributeException();
                     }
-
-                    seenPreferred |= isPreferred;
+                    pereferredConstructor = constructor;
                 }
             }
-
-            if (bestLength == -1)
+            if (pereferredConstructor is not null)
             {
-                string? message = $"A suitable constructor for type '{instanceType}' could not be located. Ensure the type is concrete and all parameters of a public constructor are either registered as services or passed as arguments. Also ensure no extraneous arguments are provided.";
-                throw new InvalidOperationException(message);
+                var matcher = new ConstructorMatcher(pereferredConstructor);
+                if (matcher.Match(parameters) == -1)
+                {
+                    ThrowMarkedCtorDoesNotTakeAllProvidedArguments();
+                }
+                return matcher.CreateInstance(provider, throwIfFailed: true)!;
             }
-
-            return bestMatcher.CreateInstance(provider);
+            var matchers = new ConstructorMatcher[constructors.Length];
+            for (int i = 0; i < constructors.Length; i++)
+            {
+                ConstructorInfo constructor = constructors[i];
+                var matcher = new ConstructorMatcher(constructor);
+                _ = matcher.Match(parameters);
```

Comment:
Might as well make the `Match` method return `void` since no one is consuming the return value anymore.